vhdbuilder/packer/test/pam/test_pam.py

```diff
@@ -0,0 +1,492 @@
+    try:
+        # child.expect_optional_sudo_auth()
+        child.expect_password_prompt(user.pw)
+        child.expect([r"\$"])
+        child.sendline("whoami")
+        child.expect(user.name)
+        print("\n-----logged in as " + user.name)
+        child.sendline("passwd")
+        print("\n-----sending current password")
+        child.expect_password_prompt(user.pw)
+        print("\n-----changing password")
+        child.expect(["[Nn]ew.*password:"])
+        child.sendline(new_pw)
+        print("\n-----sent new password")
+        i = child.expect([r"BAD PASSWORD: (.*?)(?=\n)",
+                          "Retype new.*password:"])
+        print(f"\n-----matched {i}")
+        if i == 0:
+            m = child.match.group(1)
+            print("\n-----Bad password: " + m)
+            result = False
+        elif i == 1:
+            print("\n-----retype new password")
+            child.sendline(new_pw)
+            i = child.expect([r"passwd:.*updated successfully",
+                              r"already used"])
+            if i == 0:
+                print("\n-----expecting EOF")
+                child.expect([r"\$ ", pexpect.EOF])
+                user.pw = new_pw
+                print(f"\n-----Password for {user.name} changed to {new_pw}")
+                result = True
+            elif i == 1:
+                print("\n-----Password already used")
+                result = False
+    except Exception as e:
+        print("\n-----Exception: Failed to change password: " + str(e))
+        result = False
+    finally:
+        child.close()
+
+    return result
+
+
+def change_password_w_retries(user, bad_pw):
+    """Change password, retrying until retry limit is reached.
+
+    Given a password that should be rejected, repeatedly attempt to
+    change the password for the given user.  If the password is rejected,
+    retry until the retry limit is reached, or the password is successfully
+    changed.
+
+    Keyword arguments:
+    user -- User object;
+    bad_pw -- invalid password to use
+
+    Returns:
+    Tuple of (complete, count) where:
+      complete is True if retry limit was reached, False otherwise;
+      count is the number of attempts made
+    """
+
+    print(f"\n===Attempting to change password for {user.name} to invalid " +
+          f"pw '{bad_pw}'")
+    child = BashSpawn(f"sudo login {user.name}", encoding='utf-8')
+    count = 0
+    try:
+        # child.expect_optional_sudo_auth()
+        child.expect_password_prompt(user.pw)
+        child.expect([r"\$"])
+
+        # verify successful login
+        child.sendline("whoami")
+        child.expect(user.name)
+        print("\n-----logged in as " + user.name)
+
+        # attempt to change password
+        child.sendline("passwd")
+        print("\n-----sending current password")
+        child.expect_password_prompt(user.pw)
+
+        done = False
+        while not done:
+            i = child.expect(["[Nn]ew.*password:",
```

Comment:
I think this will raise an exception if the password is ever changed -- is that intentional?

Reply:
It's only ever used in an "expected to fail" context, so yes.